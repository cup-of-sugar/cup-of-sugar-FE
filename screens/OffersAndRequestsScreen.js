import * as React from "react";
import { StyleSheet, Text, View, TouchableOpacity } from "react-native";
import Colors from "../constants/Colors";
import { ScrollView } from "react-native-gesture-handler";
import { gql } from "apollo-boost";
import { useQuery } from "@apollo/react-hooks";
import { useMutation } from "@apollo/react-hooks";

export const REQUESTS = gql`
  query ItemsUserLookingToBorrow($userId: ID!) {
    itemsUserLookingToBorrow(userId: $userId) {
      id
      name
      quantity
      available
      description
      measurement
      timeDuration
      posting {
        title
      }
      category {
        name
      }
    }
  }
`;

export const OFFERS = gql`
  query ItemsUserOfferedToLend($userId: ID!) {
    itemsUserOfferedToLend(userId: $userId) {
      id
      name
      quantity
      available
      description
      measurement
      timeDuration
      posting {
        title
      }
      category {
        name
      }
    }
  }
`;

export default function OffersAndRequestsScreen(props) {
  const navigation = props.navigation;
  const action = props.route.params.action;
  const userId = props.route.params.userId;

  if (action === "borrow") {
    const { loading, error, data } = useQuery(REQUESTS, {
      variables: {
        userId: userId
      }
    });

    if (loading) {
      return <Text style={styles.loadingText}>Loading...</Text>;
    }

    if (error) {
      return <Text style={styles.errorText}>No items found!</Text>;
    }

    if (data) {
      return (
        <View style={styles.container}>
          <Text style={styles.itemsMessage}>Items I've Requested:</Text>
          <ScrollView
            scrollIndicatorInsets={{ right: 1 }}
            style={styles.itemsContainer}
          >
            {data.itemsUserLookingToBorrow.length ? (
              data.itemsUserLookingToBorrow.map(item => {
                return (
                  <View style={styles.item} key={item.id + item.name}>
                    <Text style={styles.itemName}>
                      {item.name.toLowerCase()}
                    </Text>
                    <Text style={styles.itemName}>
                      {item.measurement
                        ? item.quantity + " " + item.measurement
                        : item.quantity && item.timeDuration
                        ? item.timeDuration
                        : item.quantity}
                    </Text>
                  </View>
                );
              })
            ) : (
              <Text style={styles.errorText}>No Requests Found</Text>
            )}
          </ScrollView>
        </View>
      );
    }
  } else if (action === "lend") {
    const { loading, error, data } = useQuery(OFFERS, {
      variables: {
        userId: userId
      }
    });
    let id;

    const DELETE_ITEM = gql`
      mutation DeletePosting($id: ID!) {
        posting: deletePosting(input: { id: $id }) {
          title
        }
      }
    `;

    let [deleteItem] = useMutation(DELETE_ITEM, {
      variables: {
        id
      },
      refetchQueries: () => [
        {
          query: OFFERS,
          variables: {
            userId: userId
          }
        }
      ]
    });

    const handleDelete = postingId => {
      deleteItem({
        variables: {
          id: postingId
        }
      }).catch(error => console.log(error));
    };

    if (loading) {
      return <Text style={styles.loadingText}>Loading...</Text>;
    }

    if (error) {
      return <Text style={styles.errorText}>No items found!</Text>;
    }

    if (data) {
      return (
        <View style={styles.container}>
          <Text style={styles.itemsMessage}>
            Items you have offered to lend that are not currently being
            borrowed:
          </Text>
          <ScrollView
            scrollIndicatorInsets={{ right: 1 }}
            style={styles.itemsContainer}
          >
            {data.itemsUserOfferedToLend.length ? (
              data.itemsUserOfferedToLend.map(item => {
                return item.available ? (
                  <View style={styles.item} key={item.id + item.name}>
                    <Text style={styles.itemName}>
                      {item.name.toLowerCase()},{" "}
                      {item.measurement
                        ? item.quantity + " " + item.measurement
                        : item.quantity && item.timeDuration
                        ? item.timeDuration
                        : item.timeDuration || item.quantity}
                    </Text>
                    <TouchableOpacity
                      style={styles.messageButton}
                      onPress={() => handleDelete(item.postingId || "1")}
                    >
                      <Text style={styles.messageButtonText}>Delete</Text>
                    </TouchableOpacity>
                  </View>
                ) : null;
              })
            ) : (
              <Text style={styles.errorText}>No Items Found!</Text>
            )}
          </ScrollView>
        </View>
      );
    }
  } else {
    return <Text style={styles.errorText}>No Items Found!</Text>;
  }
}

const styles = StyleSheet.create({
  errorText: {
    alignSelf: "center",
    color: "red",
    fontSize: 20,
    fontWeight: "bold",
    margin: 10
  },
  container: {
    flex: 1,
    backgroundColor: "#fff",
    paddingTop: 20
  },
  itemsContainer: {
    flex: 1
  },
  itemsMessage: {
    alignSelf: "center",
    fontSize: 22,
    fontWeight: "bold",
    color: "black",
    lineHeight: 35,
    margin: 20,
    textAlign: "center",
    width: 380
  },
  item: {
    alignSelf: "center",
    alignItems: "center",
    justifyContent: "space-between",
    backgroundColor: Colors.aqua,
    borderRadius: 7,
    flex: 1,
    flexDirection: "row",
    margin: 10,
    padding: 15,
    height: 70,
    overflow: "hidden",
    width: 370
  },
  itemName: {
    fontSize: 23,
    fontWeight: "bold",
    color: Colors.darkBlue,
    textAlign: "left",
    overflow: "hidden",
    width: 220
  },
  messageButton: {
    backgroundColor: Colors.darkBlue,
    borderRadius: 10,
    borderWidth: 1,
    borderColor: Colors.darkBlue,
    color: "#fff",
    height: 50,
    paddingTop: 12,
    width: 100
  },
  messageButtonText: {
    fontSize: 18,
    fontWeight: "bold",
    color: "#fff",
    textAlign: "center"
  },
  loadingText: {
    alignSelf: "center",
    fontSize: 20,
    fontWeight: "bold"
  }
});
