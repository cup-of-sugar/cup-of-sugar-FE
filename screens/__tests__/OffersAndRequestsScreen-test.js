import React from "react";
import OffersAndRequestsScreen, {
  REQUESTS,
  OFFERS,
} from "../OffersAndRequestsScreen";
import { create, act } from "react-test-renderer";

import { MockedProvider } from "@apollo/react-testing";

test("should render component", async () => {
  const mocks = [
    {
      request: {
        query: REQUESTS,
      },
      result: {
        data: {
          itemsUserLookingToBorrow: [
            {
              __typename: "ItemType",
              id: 1,
              name: "butter",
              measurement: 2,
              quantity: "sticks",
              timeDuration: null,
              category: {
                __typename: "CategoryType",
                name: "food",
              },
              description: "",
              available: true,
              posting: {
                __typename: "PostingType",
                title: "Gimme butter",
              },
            },
          ],
        },
      },
    },
  ];
  const props = {
    navigation: { navigate: jest.fn() },
    route: { params: { action: "borrow" } },
  };
  let tree;
  await act(async () => {
    tree = await create(
      <MockedProvider mocks={mocks}>
        <OffersAndRequestsScreen {...props} />
      </MockedProvider>
    );
  });
  expect(tree).toMatchSnapshot();
});

test("should render component", async () => {
  const mocks = [
    {
      request: {
        query: OFFERS,
      },
      result: {
        data: {
          itemsUserOfferedToLend: [
            {
              __typename: "ItemType",
              id: 1,
              name: "butter",
              measurement: 2,
              quantity: "sticks",
              timeDuration: null,
              category: {
                __typename: "CategoryType",
                name: "food",
              },
              description: "",
              available: true,
              posting: {
                title: "Gimme butter",
                __typename: "PostingType",
              },
            },
          ],
        },
      },
    },
  ];
  const props = {
    navigation: { navigate: jest.fn() },
    route: { params: { action: "lend" } },
  };
  let tree;
  await act(async () => {
    tree = await create(
      <MockedProvider mocks={mocks}>
        <OffersAndRequestsScreen {...props} />
      </MockedProvider>
    );
  });
  expect(tree).toMatchSnapshot();
});
