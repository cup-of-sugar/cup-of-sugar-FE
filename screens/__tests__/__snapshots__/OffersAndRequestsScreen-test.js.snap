// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component 1`] = `
<Text
  style={
    Object {
      "alignSelf": "center",
      "fontSize": 20,
      "fontWeight": "bold",
    }
  }
>
  Loading...
</Text>
`;

exports[`should render component 2`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "paddingTop": 20,
    }
  }
>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "color": "black",
        "fontSize": 22,
        "fontWeight": "bold",
        "lineHeight": 35,
        "margin": 20,
        "textAlign": "center",
        "width": 380,
      }
    }
  >
    Items you have offered to lend that are not currently being borrowed:
  </Text>
  <RCTScrollView
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    scrollIndicatorInsets={
      Object {
        "right": 1,
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#0CE1C3",
            "borderRadius": 7,
            "flex": 1,
            "flexDirection": "row",
            "height": 70,
            "justifyContent": "space-between",
            "margin": 10,
            "overflow": "hidden",
            "padding": 15,
            "width": 370,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#385A94",
              "fontSize": 23,
              "fontWeight": "bold",
              "overflow": "hidden",
              "textAlign": "left",
              "width": 220,
            }
          }
        >
          butter
          ,
           
          sticks 2
        </Text>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#385A94",
              "borderColor": "#385A94",
              "borderRadius": 10,
              "borderWidth": 1,
              "color": "#fff",
              "height": 50,
              "opacity": 1,
              "paddingTop": 12,
              "width": 100,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Delete
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
